### PR TITLE
Custom HTML5 Web Components and value attribute

### PR DESCRIPTION
**Pull request by lucianoiam**
I created this components [library](https://github.com/lucianoiam/guinda). The components support a value attribute like a standard `<input>`. Seems like LemonadeJS is giving some special treatment to this attribute and my custom components always render an empty string as the value instead of the actual value. For example:

`<g-knob value="{{self.gainValue}}"></g-knob>`

`self.gainValue = 0.25`

Always renders to:

`<g-knob value=""></g-knob>`

This patch fixes the issue by replicating the fallback code path in setAttribute() when the attribute key is 'value' and the element is neither a select, checkbox, radio, etc.

### CODE REVIEW

**Comment by hodeware** [comment]
Unfortunately, this PR will crash several components and cannot be accepted. We need to find an alternative solution for that.

**Comment by lucianoiam** [comment]
What about allowing to disable the automatic processing of attribute value on a component basis? so it gets treated just like any other attribute. My knowledge of LemonadeJS is very limited, not sure I can provide further ideas for a general solution.

**Comment by hodeware** [comment]
@lucianoiam value is a very special attribute, for example is the basis of all @bind attribute, which should work for real HTML and lemonade HTML components. I don't know exactly what you trying to archieve, but one way to handler value in a custom way:

You can do something like that.
https://jsfiddle.net/p60xvkaz/10/

**Comment by lucianoiam** [comment]
Thanks for the example.

I revisited my widgets code and they already react to changes to the value property. So in theory this line in LemonadeJS should work without the proposed patch: https://github.com/lemonadejs/lemonadejs/blob/496ad4a6c446aa7637dfdf30877d0818dad15c6e/dist/lemonade.js#L143

Is it correct to conclude custom HTML5 must not define their own value property because that is incompatible with LemonadeJS?

Trying to understand whether I hit an edge case or there is something inherently wrong with my widgets

**Comment by lucianoiam** [comment]
This is how I wrapped one of the custom elements:

```JS

export class GKnob extends lemonade.component {

    render() {
        return `<g-knob
                    min="{{self.min}}"
                    max="{{self.max}}"
                    value="{{self.value}}"
                    oninput="self.value=this.value">
                </g-knob>`;
    }

    constructor() {
        super();

        this.min = 0;
        this.max = 1;
        this.value = 0;
    }

}
```

**Comment by hodeware** [comment]
Your component should react to that the same way **input.value = 123**
yourcomponent.value = 123 - should present the value 123 wherever needs to be.

LemonadeJS only set the value of your component, internally your component should handle that.

**Comment by hodeware** [comment]
This part can be replace for @bind
```
                value="{{self.value}}"
                oninput="self.value=this.value">
```

    render() {
        return `<g-knob
                    min="{{self.min}}"
                    max="{{self.max}}"
                    @bind="self.value">
                </g-knob>`;
    }

**Comment by lucianoiam** [comment]
Components already react to changes in the value property, please see last line in https://jsfiddle.net/Lsyef6tp/1/

How does @bind now it should attach a listener to a custom input change event? my library emits these:

```
        const ev = new Event('input');
        ev.value = val;
        this.dispatchEvent(ev);
```

Note I deliberately diverged from the stock HTML elements for some design choices.

**Comment by hodeware** [comment]
@bind will syncronize the value of your component with the self property inside LemonadeJS component. It pretty much does what you have done using oninput. @bind will do two-way binding.... if your self.value changes it apply that to your component, and if you component changes it will change self.value.

**Comment by hodeware** [comment]
I have created many custom elements which implements element.val = function() {} 


That is not the only way to do it and probably what you are doing would be fine. But val() helps you a lot if you want custom two way binding with LemonadeJS using the value property.

**Comment by lucianoiam** [comment]
> @Bind will syncronize the value of your component with the self property inside LemonadeJS component. It pretty much does what you have done using oninput. @Bind will do two-way binding.... if your self.value changes it apply that to your component, and if you component changes it will change self.value.

Makes sense, the custom component self-updates its value property on user input. I confirm @bind is equivalent to specifying the value and oninput properties in the example above.

However why the unpatched LemonadeJS is not working remains a mystery, as shown in the JSFiddle the knob correctly updates when setting its value property.

I will prepare a bare-bones custom HTML5 component example that reacts to value, to rule out any issues on my side.

**Comment by hodeware** [comment]
Thanks for sharing your example, I will resolve your mistery... and yes it is a mistery indeed.

**Comment by hodeware** [comment]
One problem in your componet.

    const knob = document.createElement('g-knob');
    knob.value = 0.33
    document.body.appendChild(knob);

If you do that knob won't have any value.

**Comment by lucianoiam** [comment]
Good catch! fixed in https://github.com/lucianoiam/guinda/commit/f79e395f55da70719646d240ff0f5903b1bab0ce, thanks a lot.

The component value was being overwritten with a default when attaching the component to DOM.

There is something else going on though, the value setter is never called by LemonadeJS. Is it possible that the library expects value to be a property and not a get/set pair? the relevant component code looks like this:

```JS

    get value() {
        return this._value;
    }

    set value(newValue) {
        console.log('Never called');

        this._value = newValue;
        this._valueUpdated();
        this._dispatchSetValueEvent(this.value);
    }
```

**Comment by lucianoiam** [comment]
Trying to understand why `set value()` never gets called and at the same time no errors are printed to the console.

I inserted some debug messages here https://github.com/lemonadejs/lemonadejs/blob/496ad4a6c446aa7637dfdf30877d0818dad15c6e/dist/lemonade.js#L143

```JS
                const dummy = document.createElement('g-knob');
                console.log(dummy);
                console.log(typeof dummy);
                console.log(Object.keys(dummy));

                console.log(e);
                console.log(typeof e);
                console.log(Object.keys(e));

                e.value = v;
```

I fail to understand the output, I expected `e` to be an instance of the same type to the one returned by `document.createElement()`. First group of `console.log()` calls print:

Screenshot 2022-12-26 at 18 23 38

And the second group:

Screenshot 2022-12-26 at 18 23 50

**Comment by hodeware** [comment]
I have spent a good chunk of my day to isolate the problem outside the LemonadeJS scope. So, it seems there is a problem on your component.

This will work.
```
    let div = document.createElement('template');
    div.innerHTML = `<g-knob
                    min="0"
                    max="1" value="0.2">
                </g-knob>`;
    let test = div.content.childNodes[0];
    root1.appendChild(test);
    test.value = 0.5
```


This won't work.
```
    let div = document.createElement('template');
    div.innerHTML = `<g-knob
                    min="0"
                    max="1" value="0.2">
                </g-knob>`;
    let test = div.content.childNodes[0];
    test.value = 0.1
    root1.appendChild(test);
    test.value = 0.5
```

Notice that the difference is only if I try to define a value before appending the ELEMENT to the DOM. If I define the value before the component goes to the DOM somehow setter and getter are destroyed.

Here is an example without LemonadeJS.
https://jsfiddle.net/wnbzkp9g/37/

**Comment by lucianoiam** [comment]
As mentioned above, that bug was fixed (thanks again for pointing it out). Still, that bug should not have prevented things from working after the element is attached, only break the initial value.

The `el.value` call is failing and that is the culprit. Cannot explain why though, could you provide some testing tips?

> somehow setter and getter are destroyed.
I do not understand this part, is it something my component does wrong?

**Comment by hodeware** [comment]
This example does not use LEMONADEJS. If you make this example work, them lemonadeJS will work.
https://jsfiddle.net/wnbzkp9g/37/

**Comment by lucianoiam** [comment]
hmm while I do not discard an issue on this side, it does not seem trivial. Only the template way of doing it as shown in the JSFiddle is failing, so until now we were looking at different results (I was testing with no. 3 below) 

    const html = `<g-knob
                    min="0"
                    max="1" value="0.2">
                </g-knob>`;

Fails ( same as previous JSFiddle )
```JS
    let template = document.createElement('template');
    template.innerHTML = html;
    let el_1 = template.content.childNodes[0];
    el_1.value = 0.1
    document.body.appendChild(el_1);
    el_1.value = 0.5
```

Works ( alternate way for template )
```JS
    let el_2 = document.createRange().createContextualFragment(html).firstChild;
    el_2.value = 0.1
    document.body.appendChild(el_2);
    el_2.value = 0.5
```

Works ( no template )
```JS
    let el_3 = document.createElement('g-knob');
    el_3.value = 0.1
    document.body.appendChild(el_3);
    el_3.value = 0.5
```

**Comment by hodeware** [comment]
I can imagine, and those things are complicate. The problem LemonadeJS uses template, and the template is important since it will work with elements such as <table>, <tbody>. I am not sure why template is not working, this will require a long analysis of your script but unfortunately I won't be able to help from here.

**Comment by lucianoiam** [comment]
No need for analyzing my script, nor I think it is an issue in LemonadeJS at this point. This barebones example displays the same behavior https://jsfiddle.net/h38md92L/ (note: it prints to the JSFiddle console).

**Comment by hodeware** [comment]
Your code fails on templates which is the bases of LemonadeJS. If your code does not support template won't support LemonadeJS.
```
   let template = document.createElement('template');
    template.innerHTML = html;
    let el_1 = template.content.childNodes[0];
    el_1.value = 0.1
    document.body.appendChild(el_1);
    el_1.value = 0.5
```

**Comment by lucianoiam** [comment]
Have you checked the JSFiddle?

**Comment by lucianoiam** [comment]
It basically defines an empty HelloComponent custom component with a get/set property. Instantiating it using a template will cause the value setter to never be called. Just for fun tried to set the property name to something else and also does not work.

There is a potential risk of unrelated custom HTML5 components failing to work with LemonadeJS because of this behavior, for which I can only provide evidence but there is no explanation yet. I think it is safe to rule out issues in my library at this point too, unless there is also something substantially wrong in the hello example (also a possibility).

I am happy to perform further testing if needed

**Comment by hodeware** [comment]
I have spent a lot of time on this issue. I tried so many different possibilities and there is no much alternatives for the value, and it seems there is a bug with the template which is not related to the use of LemonadeJS, but unfortunatelly won't be a easy task to change that from a template o another possibility and don't break thousands of applications around..... If you wish to use lemonadejs one alternative is:

```
class HelloElement extends HTMLElement {

    constructor(opt) {
        super();
    }

    connectedCallback() {
        var o = this;
        Object.defineProperty(this, 'value', {
            get: function () {
                return o._value;
            },
            set: function(v) {
                o._value = v;
            }
        });
    }
}
```

**Comment by lucianoiam** [comment]
OK, it is enough to know that LemonadeJS is currently not suitable for use with arbitrary HTML5 Web Components, nor there is an issue in my own library.

I don't agree on imposing constraints onto 3rd party components as the solution though. Maybe PR #33 can serve as the starting point of a general solution, at least we now know why things are breaking.

Meanwhile I will continue working with a custom fork. Thanks for your time on this.

**Comment by hodeware** [comment]
I found the solution. The problem is:

"From the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set): It is not possible to simultaneously have a setter on a property that holds an actual value. That is why delete in the given answer is required"

This is an working example.
https://jsfiddle.net/gjxwnvdm/19/

You will need to use delete in case a property is existing, thats why if value applied before getter and setter is not working.

Here is more information:
https://stackoverflow.com/questions/67313182/webcomponents-property-setter-not-trigger-if-property-defined-before


So, if you wish to use LemonadeJS without customizations, you need to use delete this.value on the constructor.

**Comment by lucianoiam** [comment]
In the case of the element created from a template, the first setter call never fires (ie, before attaching to the DOM). There is a missing `0 : set value = 0.2` line in the JSFiddle  output:

```
"Created 0"
"0 : set value = 0.5"
"Created 1"
"1 : set value = 0.2"
"1 : set value = 0.6"
```

**Comment by hodeware** [comment]
It is not fired, but you can save the value on the constructor before you delete it as shown on the jsffidle.

**Comment by hodeware** [comment]
This total makes sense. From now, you can decide the best course of action for you.
image

Good luck on your project.

**Comment by lucianoiam** [comment]
Judging by the output of the JSFiddle with the big arrow, the initial value of the component will not be correctly set by LemonadeJS because the component is still not attached to the DOM when the first `el.value = ...` call takes place.

Just tried adding `delete this.value` to my widget, made sure LemonadeJS is unmodified 2.8.0 and as expected, the initial component value was not correctly set.

To make it clear, `delete this.value` is still NOT a solution.

**Comment by hodeware** [comment]
The biggest problem is resolved by delete. I am not sure if face other problems.

So, integrate the webcomponent with LemonadeJS and value without any problem.
https://jsfiddle.net/10bdw3m6/29/

**Comment by lucianoiam** [comment]
Removing the `delete this.value` from this new JSFiddle still allows the reactive view to work properly

**Comment by hodeware** [comment]
Luciano,
If I remove this.value the value changes but the input range keeps still...

**Comment by lucianoiam** [comment]
Fair enough, just realized there is a button so the test was incomplete. Now trying to replicate the fix in my own widget...

**Comment by hodeware** [comment]
Cool, I hope this fixes your problem.

**Comment by lucianoiam** [comment]
It did indeed, thanks a lot. I likely did something wrong earlier today, those widgets are implemented with a somehow complex class hierarchy. It is great to have isolated test code and evidence now. I honestly did not know about deleting properties before setting custom accessors, I got this in one of the notification emails from you:

_delete a property to use getter and setter is mandatory in all research I have done._

Is that the case? probably explains why I try to avoid JavaScript unless working on web stuff, lol. As mentioned earlier this wasn't trivial.

Also unrelated but I was a bit suspicious about the discussion at first, starting projects that rely on smaller unknown dependencies is risky (same goes for anyone considering my own stuff). I guess one of the first steps is finding out how maintainers deal with feedback, then you know whether it is worth. Keep up with the great work.

**Comment by hodeware** [comment]
Thanks great news! Quite happy you find the solution... yes, your case was a bit complicated so isolating the problem was crucial to find the solution. In the end everything makes sense and it was a good DOM exercise ;)

Thanks a lot. We have some other components... that might be usesul one day...
https://jspreadsheet.com/v9

**Comment by lucianoiam** [comment]
Still trying to understand why `delete.value` is needed so other authors are not misguided. I am not particularly proficient in JavaScript so take this with a grain of salt...

So far it does not look as a bug in my library, but a requirement of LemonadeJS. As you mentioned earlier `self.value` is a _very special_ property, and I think the fact that makes it special is that the library appends the `value` property to every created element. So, when a Web Component (WC) implementation also happens to implement a get/set pair for `value`, boom. As pointed out by MDN in these cases the property needs to be deleted...

This looks really more like your library making an assumption, unless it is standard practice every time an accessor is defined, its matching property is deleted for safety (I lack the experience here). The assumption can be turned into a requirement with a bit of proper documentation, which is great to know is on the way.

I give you credit for spotting the bug in guinda that overwrited the existing `value` with default values if the property was written before attaching the element to the DOM. Again an unfortunate coincidence that masked the root issue (colliding property and accessor). Adding insult to injury, I made some mistakes while testing your JSFiddles.

In conclusion one bug on my side was fixed, and the second update was merely for compatibility with LemonadeJS. I am not sure this is possible at all but ideally user elements should remain untouched and any unsurmountable exception clearly documented.

**Comment by hodeware** [comment]
That is not actually related to LemonadeJS in particular, and you would probably bump on that one day or another since you are using getters and setters. But, this is not only related to the property value, could be related to any conflict between a property and getters/setters with the same name. One way to understand that is....



let obj = {
    value: 123,
}

If you wish to change obj.value = 567 is fine, you are changing the property directly.

But, if you are using getters and setters, you cannot have a property value

let obj {
    get value()
    set value()
    value: 123// this will cause issues 
}

Thats why you have obj._value = 123

I share with you some documentation from Mozilla yestarday. I belive you can add something to the __construct of your classes so your users don't need to do that. You pretty much need to do the delete for all properties that has getters and setters with the same name.

Hope that is clear...

**Comment by lucianoiam** [comment]
The documentation from Mozilla is helpful but has nothing to do with what I am stating here, as the widgets library does not define properties and accessors with the same name simultaneously. That explains why it has worked for some time in other environments.

I am thinking about a responsibility issue. Are developers expected to _always_ delete properties when implementing accessors? that is what you seem to imply, but not what the MDN explicitly states. In other words, is it common practice to delete properties _just for safety_?

I fail to understand who is responsible of taking care of this situation:

1. Code which implements accessors (because the delete thing is documented, or common practice).
2. Code which sets properties to an instance potentially implementing accessors with the same name (in that case that must be documented).

I do not want to turn this into a philosophical debate as I lack the knowledge and experience, just pointing out I'm skeptical I have a bug on this front. And deleting every single accessor just in case, sounds horrible but could be the case, IMO JavaScript is not particularly beautiful after all 😅

**Comment by hodeware** [comment]
You only need to delete if you have a getter or setter with the same name.

**Comment by lucianoiam** [comment]
> You only need to delete if you have a getter or setter with the same name.

Taking this literally, then my library was just fine. The issue arises only when LemonadeJS gets involved, it creates a new compatibility requirement that must be documented. We cannot just expect 3rd party existing Web Components libraries to already call `delete this.value`. 

It is totally fine if LemonadeJS has some minor requirements to keep things simple, just have them clearly documented. It is a responsibility issue (in the software sense).

**Comment by hodeware** [comment]
What is happening with LemonadeLS.

1) The object is create
2) LemonadeJS set the property based on the attributes defined in the tag
3) Them you getter and setters are created
4) Append the component to the DOM

Because LemonadeJS define property before the getter and setters are created, them you have a conflict.

**Comment by lucianoiam** [comment]
I would say this one is key:

> 2. LemonadeJS set the property based on the attributes defined in the tag

Regarding `<template>`, again I am not very proficient but know there is `DOMParser` and `createContextualFragment()`. As shown earlier, the latter comes with its own quirks...

**Comment by hodeware** [comment]
Yes, when you define a attribute in the tab <Test value="{{self.value}}" />, the value will be defined to the Test object. But, test object is receiving getters and setters delayd due the template engine. 

We have tried a lot of different options for the template in the best, template is the most flexible. We will re-check other possibilities for version 3.
